vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -238,15 +238,36 @@ import { isFocusable } from '@vaadin/grid/src/vaadin-grid-active-item-mixin.js';
         });
         grid._createPropertyObserver('activeItem', '__activeItemChangedDetails', true);
 
-        grid.$connector._getPageIfSameLevel = tryCatchWrapper(function (parentKey, index, defaultPage) {
-          let cacheAndIndex = grid._cache.getCacheAndIndex(index);
-          let parentItem = cacheAndIndex.cache.parentItem;
-          let parentKeyOfIndex = parentItem ? grid.getItemId(parentItem) : root;
-          if (parentKey !== parentKeyOfIndex) {
-            return defaultPage;
-          } else {
+        grid.$connector._getSameLevelPage = tryCatchWrapper(function (parentKey, index, defaultPage) {
+          const cacheAndIndex = grid._cache.getCacheAndIndex(index);
+          const parentItem = cacheAndIndex.cache.parentItem;
+          const parentKeyOfIndex = parentItem ? grid.getItemId(parentItem) : root;
+          if (parentKeyOfIndex === parentKey) {
             return grid._getPageForIndex(cacheAndIndex.scaledIndex);
           }
+          // Try to match level by going up in hierarchy. This accounts for the cases where
+          // the parent of an item is not visible, but has the same depth as of the request.
+          // This prevents clearing the cache for those visible items, which can create a loop.
+          return this._getSameLevelPageByHierarchy(parentKey, defaultPage, parentKeyOfIndex);
+        });
+
+        grid.$connector._getSameLevelPageByHierarchy = tryCatchWrapper(function (parentKey, defaultPage, parentKeyOfIndex) {
+          // Root item was reached
+          if (parentKeyOfIndex === root) {
+            return defaultPage;
+          }
+          // There is no parent item to match level in cache
+          const parentCacheAndIndex = grid._cache.getCacheAndIndexByKey(parentKeyOfIndex);
```

Comment:
`cache.getCacheAndIndexByKey` is an expensive operation because it traverses the tree downward. I would recommend using the `cache.parentCache` reference instead.

Reply:
I updated the logic. The initial version called `getCacheAndIndex` twice regardless. Now it is called once at the start to get an initial cache. Then we traverse the tree upwards using `cache.parentCache`. If a match is found, `cache.getCacheAndIndexByKey` is called once to determine the page of the same level index.